### ISSUE TITLE
[Android] Throws runtime error in `mscorlib`, expected class `System.Runtime.CompilerServices.ValueTaskAwaiter1`

### ISSUE THREAD

**Issue by sacOO7** [dotnet/runtime#82193]
### Description

I am unable to use [ably package](https://www.nuget.org/packages/ably.io) with a dependency on `System.Threading.Channels` (the error might also have origin from `System.Threading.Tasks.Extensions`/`System.Runtime.CompilerServices.Unsafe` assembly) , throws below runtime exception on MAUI android.

Throws a runtime error while trying to run on android  =>

```
System.TypeLoadException: 'Could not load type of field 'IO.Ably.Realtime.Workflow.RealtimeWorkflow+d__36:<>u__2' (11) due to:
Could not resolve type with token 010000ba from typeref (expected class 'System.Runtime.CompilerServices.ValueTaskAwaiter1' 
in assembly 'mscorlib, Version=2.0.5.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e')
assembly:mscorlib, Version=2.0.5.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e type:System.Runtime.CompilerServices.ValueTaskAwaiter1 member:(null)
```

 It works on a regular net6.0 library, MAUI windows machine, and the latest Xamarin Forms.

I tried all options to disable assembly trimming as per 
https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-6-0
but the issue still persists and the app throws an error at runtime. 

I strongly feel the issue is related to a missing type in `mscorlib` same as the one mentioned in https://github.com/dotnet/maui/issues/5078 and https://github.com/dotnet/maui/issues/5142

### Steps to Reproduce

1. Clone the project from https://github.com/sac-fork/MauiAppAblyDemo
2. Build and run it on android emulater ( I used default Pixel 5 - API 33 (API 13 - API 33) )
3. Tap on `Click me` button, it will throw above error. 

### Link to public reproduction project repository

https://github.com/sac-fork/MauiAppAblyDemo

### Version with bug

6.0 / 7.0 

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android API 13 to 33

### Did you find any workaround?

No workarounds were found.

### Relevant log output

```shell
System.TypeLoadException: 'Could not load type of field 'IO.Ably.Realtime.Workflow.RealtimeWorkflow+d__36:<>u__2' (11) due to: Could not resolve type with token 010000ba from typeref (expected class 'System.Runtime.CompilerServices.ValueTaskAwaiter1' in assembly 'mscorlib, Version=2.0.5.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e') assembly:mscorlib, Version=2.0.5.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e type:System.Runtime.CompilerServices.ValueTaskAwaiter1 member:(null)
```

**Comment by sacOO7** [comment]
@rachelkang I am not sure why this issue was transferred to `dotnet/runtime`, but I am skeptical about getting it fixed sooner. ( There are already 5k+ active issues on this repository ) Can you please add this issue to your priority queue ?

**Comment by sacOO7** [comment]
Any update on this @Redth and @XamlTest ?

**Comment by marek-safar** [comment]
The problem is very likely down to this dependency `'System.Runtime.CompilerServices.ValueTaskAwaiter1'
in assembly 'mscorlib, Version=2.0.5.0`. One of your packages is not targeting .NET5+ but original Xamarin only. This should in theory work but it's always tricky combination.

@steveisok could you check why it's failing?

**Comment by sacOO7** [comment]
@marek-safar can you try this app ->https://github.com/sac-fork/MauiAppAblyDemo?

 If you have a workaround to get this working, it will be super useful 🥇

**Comment by akoeplinger** [comment]
@sacOO7 the short answer is that you need to add a `net6.0` target to your nuget package to fix the issue.

Long answer:

.NET (Core) provides a compatibility mechanism for using assemblies which were compiled against the legacy .NET Framework assembly layout (e.g. mscorlib.dll, System.dll etc). The way this works is that we ship a tiny shim mscorlib.dll which is only used to forward the types to their new locations in a different assembly on modern .NET (e.g. System.Private.CoreLib.dll).

However this breaks down in this case since the classic Xamarin.Android frameworks used the classic Mono implementation of .NET Framework and we made a few additions that didn't exist in the original Microsoft .NET Framework implementation, mainly to support .NET Standard 2.1 use cases.

Your nuget package contains these target frameworks:
<img width="315" alt="image" src="https://user-images.githubusercontent.com/1376924/220969797-cc13a02c-7466-42a0-a4de-810f3b253136.png">

The `net6.0-android` target framework has a compatibility fallback path with `monoandroid` so the build uses the assembly from that folder instead of e.g. the `netstandard2.0`, since it is considered more compatible.
However that assembly was compiled against a Xamarin.Android-specific mscorlib.dll which contained types like `ValueTaskAwaiter<T>` and `ReadOnlySpan<T>` so the assembly expects them to be there. The compatibility shim mscorlib.dll however doesn't have type forwarders for these types (since they didn't exist on the Microsoft .NET Framework implementation) and we get the TypeLoadException.

A quick way to hack it is to go into your local nuget package cache folder `<home directory>/.nuget/packages/ably.io/1.2.10/lib/` and remove the `monodroid` folder there, then rebuild the app. That will make it pick the `netstandard` version instead which works:

<img width="1547" alt="image" src="https://user-images.githubusercontent.com/1376924/220976335-18959722-7321-486d-881f-9e9c75befb65.png">

But the real solution is to add a proper `net6.0` target to your nuget package so that assembly will be picked instead of the `monoandroid` one.

**Comment by sacOO7** [comment]
@akoeplinger Thanks for the answer!

I had one final question.
 If target for `.Net6` is not yet added, why does it show support for `.Net 6`/ `.Net7` under nuget
https://www.nuget.org/packages/ably.io#supportedframeworks-body-tab ?

**Comment by akoeplinger** [comment]
> If target for `.Net6` is not yet added, why does it show support for `.Net 6`/ `.Net7` under nuget

See the legend at the bottom of the table, that target is "computed", because we have this compatibility layer so `monoandroid` or `net46` works on modern .NET.

<img width="378" alt="image" src="https://user-images.githubusercontent.com/1376924/221002333-07b17ab8-2246-4688-bcdd-df2747926194.png">

We can't throw an error since this is supposed to work, you're just hitting an edge case where it doesn't. The compatibility layer is best effort which is why we're recommending adding a real `net6.0` target.

**Comment by sacOO7** [comment]
Got it @akoeplinger. I had few small questions coming up.

1. Can we make project choose `netstandard 2.0` assembly instead of `xamarin-android` at compile time through some configuration.
2. How can I add targets for `.Net 6` / `.Net 7` so it reflects in nuget package, any link to the related doc will be super useful?

**Comment by akoeplinger** [comment]
> 1. Can we make project choose netstandard 2.0 assembly instead of xamarin-android at compile time through some configuration.

No, the only option I can think of would be to remove the `monoandroid` folder from your nuget package so it always picks the `netstandard2.0` one, but I assume this won't work because you likely have Android-specific code otherwise the  `monoandroid` target doesn't make sense.

> 2. How can I add targets for .Net 6 / .Net 7 so it reflects in nuget package, any link to the related doc will be super useful?

That depends on how your nuget package build works today. You'd need to build with the modern .NET SDK and set `<TargetFramework>net6.0</TargetFramework>` in your csproj file.

**Comment by sacOO7** [comment]
Great, we will fix this as per your suggestion and will let you know if anything comes up !!!

Thanks again, this is really helpful : )

**Comment by akoeplinger** [comment]
I've also just experimented with adding these missing type forwarders to the mscorlib.dll shim and it fixes the issue as well, i.e. the `monoandroid` assembly works.

I opened a PR, we'll see if we can maybe backport it to 7.0: #82618

**Comment by sacOO7** [comment]
@akoeplinger wow, this is amazing : ) 

When can we expect this to get merged and available after release?

**Comment by akoeplinger** [comment]
@sacOO7 we merged the backport to the release/7.0 branch and is currently targeted for  the 7.0.5 release.

**Comment by sacOO7** [comment]
Thanks @akoeplinger .
